http/client/src/main/java/org/wildfly/security/http/client/ElytronHttpClient.java

```diff
@@ -0,0 +1,211 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.http.client;
+
+import org.wildfly.security.http.client.exception.ElytronHttpClientException;
+import org.wildfly.security.http.client.utils.DigestHttpMechanismUtil;
+import org.wildfly.security.http.client.utils.HttpMechClientConfigUtil;
+import org.wildfly.security.http.client.utils.ElytronMessages;
+
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Elytron client for HTTP authentication
+ *
+ * @author <a href="mailto:kekumar@redhat.com">Keshav Kumar</a>
+ */
+public class ElytronHttpClient {
+
+    private static final String AUTHORIZATION = "Authorization";
+    private final HttpClient client;
+    private final HttpMechClientConfigUtil httpMechClientConfigUtil;
+    private String lastURI;
+    private Map<String, String> authParams;
+    private String userName;
+    private String password;
+    private String previousMechanism;
+    private DigestHttpMechanismUtil digestHttpMechanismUtil;
+
+    public ElytronHttpClient() {
+        this.client = HttpClient.newHttpClient();
+        this.httpMechClientConfigUtil = new HttpMechClientConfigUtil();
+        authParams = new HashMap<>();
+        previousMechanism = null;
+        lastURI = null;
+        digestHttpMechanismUtil = new DigestHttpMechanismUtil();
+    }
+
+    private static String basicAuth(String username, String password) {
+        return "Basic " + Base64.getEncoder().encodeToString((username + ":" + password).getBytes());
+    }
+
+    public HttpResponse evaluateDigestMechanism(String uri) throws Exception {
+
+        HttpRequest request;
+        request = digestHttpMechanismUtil.createDigestRequest(uri,userName,password,authParams);
+
+        if(lastURI==null || !(lastURI.equals(uri))){
+            lastURI = uri;
+        }
+
+        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+
+        if (response.statusCode() == 401) {
+            String authHeader = response.headers().allValues("www-authenticate").get(0);
+            digestHttpMechanismUtil.updateAuthParams(authHeader,authParams);
+            request = digestHttpMechanismUtil.createDigestRequest(uri,userName,password,authParams);
+            response = client.send(request, HttpResponse.BodyHandlers.ofString());
+        }
+        return response;
+    }
+
+    //To test header values from ElytronHttpClientTest
```

Comment:
@keshav-725 This comment should be removed and the javadoc describing the method should be added.